### ISSUE TITLE
toBeInstanceOf does not properly recognise instance when using baseUrl in tsconfig

### ISSUE THREAD

**Issue by sosnowski** [vitest-dev/vitest#2686]
### Describe the bug

Hello All,

I've noticed unexpected behaviour that I believe is a bug. I've replicated it on Stackblitz, you can check it here: https://stackblitz.com/edit/node-qasnfa

As you can see, the tests fail on `toBeInstanceOf` check.

The tested function is very simple, it just creates an instance of the `User` class
```ts
import { User } from './user';

export const createDefaultUser = () => {
  return new User('DefaultName');
};
```

Now, this test fails, even though it should not:
```ts
import { createDefaultUser } from 'src/create';
import { User } from 'src/user';

describe('Test User class', () => {
  it('Should create user instance', () => {
    const u = createDefaultUser();
    expect(u).toBeInstanceOf(User);
  });
});
```

It fails, because `create.ts` file is importing `user.ts` "directly" `./user.ts`, while the test suite is referrng to it in relation to `tsconfig` `baseUrl` setting.

If I change the import statement in the test suite...
```ts
import { createDefaultUser } from '../src/create';
import { User } from '../src/user';

describe('Test User class', () => {
  it('Should create user instance', () => {
    const u = createDefaultUser();
    expect(u).toBeInstanceOf(User);
  });
});
```
It works perfectly

### Reproduction

https://stackblitz.com/edit/node-qasnfa

Just run `npm run test`

### System Info

```shell
This is from Stacblitz but the same issue occurs on my MacOS

  System:
    OS: Linux 5.0 undefined
    CPU: (8) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Memory: 0 Bytes / 0 Bytes
    Shell: 1.0 - /bin/jsh
  Binaries:
    Node: 16.14.2 - /usr/local/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 7.17.0 - /usr/local/bin/npm
  npmPackages:
    vitest: ^0.27.1 => 0.27.1
```

### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by sheremet-va** [comment]
Vitest doesn't recognise tsconfig. If you want to rely on it, install [vite-tsconfig-paths](https://www.npmjs.com/package/vite-tsconfig-paths) plugin.

Your particular case correctly resolves url, because it's relative to `root`. If you change `baseUrl`, it will still be resolved. Your edge case should be resolved by #2687, but if you rely on tsconfig, you should install plugin.

**Comment by sheremet-va** [comment]
This syntax is not supported without [vite-tsconfig-paths](https://www.npmjs.com/package/vite-tsconfig-paths). The fact that it "works" sometimes is a coincidence. When resolving paths, Vite thinks that `src/user` is relative to the current folder.
